File: tzio-core/spec/src/main/java/com/kineolyan/tzio/v1/api/ops/Operations.java
package com.kineolyan.tzio.v1.api.ops;

import com.kineolyan.tzio.v1.api.ref.InputReferenceType;
import com.kineolyan.tzio.v1.api.ref.OutputReferenceType;

/**
 * Public facade for creating operations.
 */
public class Operations {

	private Operations() {}

	public static OperationType MOV(final InputReferenceType input, final OutputReferenceType output) {
		return new MovOperation(input, output);
	}

	public static OperationType SAV(final int slot) {
		return new SavOperation(slot);
	}

	public static OperationType SWP(final int slot) {
		return new SwpOperation(slot);
	}

	public static OperationType ADD(final InputReferenceType input) {
		return new AddOperation(input);
	}

	public static OperationType SUB(final InputReferenceType input) {
		return new SubOperation(input);
	}

	public static OperationType NEG() {
		return new NegOperation();
	}

	public static OperationType LABEL(final String label) {
		return new LabelOperation(label);
	}

	public static OperationType JMP(final String label) {
		return new JmpOperation(label);
	}

	public static OperationType JEZ(final String label) {
		return new JezOperation(label);
	}

	public static OperationType JNZ(final String label) {
		return new JnzOperation(label);
	}

	public static OperationType JLZ(final String label) {
		return new JlzOperation(label);
	}

	public static OperationType JGZ(final String label) {
		return new JgzOperation(label);
	}

	public static OperationType JRO(final InputReferenceType input) {
		return new JroOperation(input);
	}

}

File: tzio-core/clj/src/clj/ops/operations.clj
(ns clj.ops.operations
  (:require [clj.refs.references :as refs])
  (:import 
    (com.kineolyan.tzio.v1.api.ops
        MovOperation
        SavOperation
        SwpOperation 
        AddOperation 
        SubOperation 
        NegOperation 
        LabelOperation 
        JmpOperation 
        JezOperation 
        JnzOperation 
        JlzOperation 
        JgzOperation 
        JroOperation)))
  
(defn mov
  [^MovOperation type]
  [:mov (refs/convert-input (.-input type)) (refs/convert-output (.-output type))])

(defn sav
  [^SavOperation type]
  [:sav (.-slot type)])

(def swp
  [^SwpOperation type]
  [:swp (.-slot type)])

(defn add
  [^AddOperation type]
  [:add (refs/convert-input (.-input type))])

(defn sub
  [^SubOperation type]
  [:sub (refs/convert-input (.-input type))])

(defn neg
  [_]
  [:neg])

(defn lbl
  [^LabelOperation type]
  [:label (.-label type)])

(defn jmp
  [^JmpOperation type]
  [:jmp (.-label type)])

(defn jez
  [^JezOperation type]
  [:jez (.-label type)])

(defn jnz
  [^JnzOperation type]
  [:jnz (.-label type)])

(defn jgz
  [^JgzOperation type]
  [:jgz (.-label type)])

(defn jlz
  [^JlzOperation type]
  [:jlz (.-label type)])

(defn jro
  [^JroOperation type]
  [:jro (refs/convert-input (.-input type))])

(defn convert
  [type]
  (neg))

(defmulti is-label? (fn [[type & remaining]] type))
(defmethod is-label? :label [& _] true)
(defmethod is-label? :default [& _] false)

(defn get-label [[_ lbl]] lbl)

(defn index
  [operations]
  (let
    [
      [ops idx] 
      (reduce 
        (fn [[ops idx] op]
          (if 
            (is-label? op)
            [
              ops
              (assoc! idx (get-label op) (count ops))]
            [
              (conj! ops op)
              idx]))
        [(transient []) (transient {})]
        operations)]
    [
      (persistent! ops)
      (persistent! idx)]))
